doc/designs/quic-design/tx-packetiser.md

```diff
@@ -0,0 +1,433 @@
+                               size_t frame_length);
+```
+
+### Stream Send Buffers
+
+Data from the stream send buffers is treated specially.  The packetiser knows
+how much space is left in each packet and it will request that amount of data
+from the stream send buffers.  The stream send buffers will return a
+constructed frame header and a pointer to the steam data and length.
+
+```c
+__owur int ossl_quic_get_app_data(SSL_QUIC_TX_PACKETISER *tx, size_t request,
+                                  const OSSL_QUIC_FRAME **frame,
+                                  const void *data, size_t data_len);
+```
+
+#### Notes
+
+* Unclear how to best free the data after sent data was acked.
+  The data will be fragments from the buffers so the stream send buffers will
+  need to remember which fragment have been sent and which are pending and
+  only free once everything is sent:
+
+```c
+int ossl_quic_free_app_data(OSSL_QUIC_FRAME *frame,
+                            void *data, size_t data_len);
+```
+
+### TLS Handshake Record Layer
+
+Uses the Record Layer API to implement the inner TLS-1.3 protocol handshake.
+It produces the QUIC crypto frames which are queued using the same mechanism
+as the [Stream Send Buffers](#stream-send-buffers) above.
+
+### Flow Controller and Statistics Collector
+
+To make decisions about what frames to coalesce, the packetiser relies
+on the flow controller to enforce stream and connection bandwidth limits
+[RFC 9000 4.1 Data Flow Control].
+
+```c
+/*
+ * Return the maximum amount of data that is permitted for the given stream.
+ * This includes both the stream limit and it's associated connection limit.
+ */
+size_t ossl_quic_flow_maximum_size(QUIC_STREAM *stream);
+
+/*
+ * Information the flow controller that an amount of data has been queued for
+ * sending to a stream.
+ */
+int ossl_quic_flow_sent_data(QUIC_STREAM *stream, size_t bytes);
+```
+
+### Congestion Controller
+
+Also part of the frame coalescing decision is the congestion controller
+[RFC 9002].  For MVP, this will be a _just send it_.
+
+```c
+/*
+ * Pluggable congestion controller APIs go here
+ * Extract that is required from #18018
+ */
+```
+
+### QUIC Write Record Layer
+
+Coalesced frames are passed to the QUIC record layer for encryption and sending.
+To send accumulated frames as packets to the QUIC Write Record Layer:
+
+```c
+__owur int ossl_quic_send_packets(SSL_QUIC_TX_PACKETISER *tx,
+                                  QUIC_CONNECTION *conn);
+```
+
+If _stream_ is NULL, some, none or all queued packets for the connection
+_tx_ is associated with will be queued for transmission, either now or later.
+
+If _stream_ is not NULL, queued data for this stream will be sent
+to the record layer immediately, subject to congestion and flow control
+limitations.
+
+The write record layer will need a call equivalent to `writev(2)` in order to
+avoid copying data needlessly.  Each packet is defined and sent to the write
+record layer in fragments:
```

Comment:
None of this section looks correct. I expect the first object here to be a QUIC write record layer. I expect to pass a `OSSL_QRL` object as the first parameter (as named in #18949). We don't yet know what the QUIC write record layer API will look like but I imagine it will be similar to the TLS one which has a proposed API like this

````
    int (*write_records)(OSSL_RECORD_LAYER *rl, OSSL_RECORD_TEMPLATE **templates,
                         size_t numtempl, size_t allowance, size_t *sent);
    int (*retry_write_records)(OSSL_RECORD_LAYER *rl, size_t allowance,
                               size_t *sent);
````

Reply:
Reworked based on current state of #18949